broadinstitute/single_cell_portal_core
# Migrating to Google Life Sciences API (SCP-5170)

#### BACKGROUND & CHANGES
This update migrates SCP off of using the Google Pipelines API (PAPI) for running file ingest jobs.  This API was never released beyond `v2alpha1` and the associated client libraries are now deprecated.  While no announcement has been made yet, it is conceivable that Google will remove this service entirely at some point.  The successor to the Pipelines API is the [Cloud Life Sciences API](https://cloud.google.com/life-sciences/docs/reference/rest).  This service is nearly identical to PAPI in terms of the interface & functionality, but is more actively maintained and has much better documentation.  Incidentally, Terra had previously switched Cromwell to use this service as well.

The migration consists mainly of updating libraries and class names, and some minor method refactoring to account for small changes in implementation or naming schemes.  Additionally, offline work has already been done in all development & staging environments to grant our service accounts the correct permissions to run pipelines (`roles/genomics.pipelinesRunner`, listed as "Cloud Life Sciences Workflows Runner" in Google IAM ), and adding in a new environment variable to store the `GOOGLE_PROJECT_NUMBER` which is required to submit/retrieve operations.

More information on the migration can be found here: https://cloud.google.com/life-sciences/docs/how-tos/migration

#### MANUAL TESTING
1. Boot as normal (including DelayedJob) and sign in
2. Go to any study and upload a new file, like the cluster file at `test/test_data/cluster_example.txt`
3. Confirm that you see the job launch in `development.log`
```
Request object sent to Google Life Sciences API, excluding 'environment' parameters:
---
:actions:
  :commands:
  - python
  - ingest_pipeline.py
  - "--study-id"
  - 63bdbf486cb53400554c7323
  - "--study-file-id"
  - '649318494041354528480037'
  - "--user-metrics-uuid"
  - 0fc4d8e5-6c4c-422e-94a3-c78801757202
  - ingest_cluster
  - "--cluster-file"
  - gs://fc-f5fd5b66-3ea7-47c6-a0e5-fd257c6565e0/cluster_example.txt
  - "--ingest-cluster"
  - "--name"
  - Cluster 1
  - "--domain-ranges"
  - "{}"
  :image_uri: gcr.io/broad-singlecellportal-staging/scp-ingest-pipeline:1.26.1
  :labels: {}
  :timeout: {}
...

Ingest run initiated: projects/342448556558/locations/us-central1/operations/4659277892119481418, queueing Ingest poller
IngestJob poller: projects/342448556558/locations/us-central1/operations/4659277892119481418 is not done; queuing check for 2023-06-21 11:34:46 -0400
```
4. Wait for the job to complete and confirm that you see the success email and that the clustering object is available to view in your study
